Attributable errors

joostjager opened the pull request:
Error attribution is important to properly penalize nodes after a payment failure occurs. The goal of the penalty is to give the next attempt a better chance at succeeding. In the happy failure flow, the sender is able to determine the origin of the failure and penalizes a single node or pair of nodes.

Unfortunately it is possible for nodes on the route to hide themselves. If they return random data as the failure message, the sender won't know where the failure happened.

This PR updates the failure encryption mechanism so that the failure source can always be determined.

References:
* https://lists.linuxfoundation.org/pipermail/lightning-dev/2022-October/003723.html
* https://github.com/lightning/bolts/pull/1044

joostjager commented on crypto.go:
```diff
@@ -292,30 +292,40 @@ func (o *OnionErrorDecrypter) DecryptError(encryptedData []byte) (
+		actualHmac := hmacs[i*sha256.Size : (i+1)*sha256.Size]
+
+		// If the hmac does not match up, exit with a nil message.
+		if !bytes.Equal(actualHmac, expectedHmac[:]) && sender == 0 {
```
Magic happening here.

joostjager committed:
sphinx: initialize NewOnionErrorEncrypter with shared secret directly

Allow for more flexible usage of the error encrypter. This is useful when upgrading an
existing legacy error encrypter to attributable errors in lnd.

joostjager committed:
sphinx: export GenerateSharedSecret

halseth commented on fat_error_decrypt.go:
```diff
@@ -0,0 +1,201 @@
+
+func (o *OnionFatErrorDecrypter) shiftHmacsLeft(hmacs []byte) {
+	if len(hmacs) != o.allHmacsLen {
+		panic("invalid hmac block length")
```
return error instead?

joostjager replied:
Check removed, because length checking is already carried out at a higher level.

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

carlaKC commented on fat_error_decrypt.go:
```diff
@@ -0,0 +1,201 @@
+		)
+		actualHmac := hmacs[i*sha256.Size : (i+1)*sha256.Size]
+
+		// If the hmac does not match up, exit with a nil message.
+		if !bytes.Equal(actualHmac, expectedHmac) && sender == 0 {
+			sender = i + 1
+			msg = nil
+		}
```
Should we break or error here? Comment says that we exit, but we'll continue on to extract the payload.

joostjager replied:
It looks a bit strange indeed, but the idea (copied from the legacy error decryption) is to decode in constant time. It seems a bit too much because there might very well be other loops in lnd over hops that aren't constant time. There is a comment right above the for loop explaining it.

halseth replied:
Agree that it's easy to miss the rationale above the for loop while reading the code here. Perhaps add a comment here as well.

joostjager replied:
Added sentence to repeat that we first finish the constant number of iterations.

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

halseth commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,103 @@
+)
+
+type AttributableErrorStructure struct {
+	HopCount        int
```
these are `u16`,`u8` etc in the BOLT, maybe use the same here?

joostjager replied:
I used ints here to prevent casting to a wider data type in the various calculations that use these values. I think you could argue that the `byte` that is specified in the bolt is a limitation on the wire level and that the code in this repo doesn't need to follow that limitation necessarily?

halseth replied:
I think it would be nice to have the types 1:1. How bad would the casting be?

Roasbeef replied:
I think the casting overhead would be negligible in terms of affected code sites. Inheriting the proper int types here aides in readability IMO. 

Also in most cases, you don't really need to cast for comparisons, as you can make generic funcs that use `constraints. Integer` or w/e.

Roasbeef replied:
Also missing `godoc` comments here. 

Is this an error struct (actually `error` type), or something else entirely? In either case, adding "Structure" to the end seems superflous.

joostjager replied:
It is not an error struct, but a description of the actual structure of the onion failure. I think the addition `Structure` is appropriate in this case. Added godoc to explain.

halseth commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+type attributableErrorBase struct {
+	maxHops             int
```
could make these returned by methods on the above `struct` instead. I think that is cleaner than pre-calculating constants.

joostjager replied:
I gave that a try, but I don't think the addition of all the parenthesis for the new function calls make the code more readable.

joostjager replied:
As you both don't seem to be too bothered with the parenthesis, I went ahead and did the conversion.

halseth commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,103 @@
+	_, _ = hash.Write(payloads[:(o.maxHops-position)*o.payloadLen])
+
+	// Include downstream hmacs.
+	var hmacsIdx = position + o.maxHops
```
if max hops is 20, and I'm at position 10, how will a hmac index of 30 be correct? Would be great with some helpful comments here.

joostjager replied:
Yes indeed, this is not easy to understand. But it is correct. Every position has one or more hmacs associated with it. So hmac index isn't staying below max hops.

halseth commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,103 @@
+			hmacs[hmacsIdx*sha256.Size : (hmacsIdx+1)*sha256.Size],
+		)
+
+		hmacsIdx += o.maxHops - j - 1
```
could you add some more comments to clarify what the various indexes mean here?

joostjager replied:
Comments and unit test added.

Roasbeef commented:
cc @bitromortac

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

joostjager commented:
!lightninglabs-deploy mute

Roasbeef commented on crypto.go:
```diff
@@ -237,8 +237,8 @@ func onionEncrypt(sharedSecret *Hash256, data []byte) []byte {
-// bytes. We then add the size of the sha256 HMAC as well.
-const minOnionErrorLength = 2 + 2 + 256 + sha256.Size
+// bytes.
+const minOnionErrorLength = 2 + 2 + 256
```
Perhaps a new constant is needed for processing w/ and w/o this new behavior? Would allow us to eliminate the `+sha256.Size` below.

Roasbeef replied:
Also IIUC, this is still valid for errors that don't include this new information.

joostjager replied:
I don't think this commit changed the logic for the legacy errors. But good suggestion to add another constant, done.

Roasbeef commented on crypto.go:
```diff
@@ -206,11 +206,11 @@ func blindBaseElement(blindingFactor btcec.ModNScalar) *btcec.PublicKey {
 
 // generateSharedSecret generates the shared secret by given ephemeral key.
-func (r *Router) generateSharedSecret(dhKey *btcec.PublicKey) (Hash256, error) {
+func (r *Router) GenerateSharedSecret(dhKey *btcec.PublicKey) (Hash256, error) {
```
Godoc above not updated. 

Is this just so the implemented interface only has public methods, or that this method is to be used elsewhere?

joostjager replied:
Godoc updated.

The method is used in #7139. Access is needed to the shared secret so that the type of onion encryptor can be 'upgraded' after reading the error structure parameters from the onion tlv.

Roasbeef commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func newAttributableErrorBase(
+	structure *AttributableErrorStructure) attributableErrorBase {
```
`onionError` instead?

Roasbeef commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,124 @@
+		// 1-byte payload type followed by the payload data.
+		payloadLen = 1 + payloadDataLen
+
+		totalHmacs = (structure.HopCount *
```
Could use a comment or spec link explaining the derivation here.

Roasbeef commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,124 @@
+
+// getMsgComponents splits a complete failure message into its components
+// without re-allocating memory.
+func (o *attributableErrorBase) getMsgComponents(data []byte) ([]byte, []byte,
```
Can use a new struct here for the message component so it's easier to read than figuring out at which index of which byte slice is needed.

Roasbeef replied:
Could also use a comment outlining the expected error/byte format.

Roasbeef commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,124 @@
+	// therefore be skipped. The first hmac to write is part of the block of
+	// hmacs that was written by the first downstream node. Which hmac exactly
+	// is determined by the assumed position of the current node.
+	var hmacIdx = maxHops + position
```
`var` here to hint that'll be modified later on in the func?

joostjager replied:
Not a conscious decision. But renamed to `currentHmacIdx` to underline that this is a working variable.

Roasbeef commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,124 @@
+	hash := hmac.New(sha256.New, umKey[:])
+
+	// Include message.
+	_, _ = hash.Write(message)
```
This is actually an encrypted message right? So the ciphertext and not the plaintext. 

If not, then we need to revise the proposal, as encrypt-then-mac is a must as we want to ensure integrity of the ciphertext. Otherwise the MAC may also give some information about the plaintext itself. 

Might just have an out of date mental model tho....decided to take a look at the code before the spec to reaload some context.

Roasbeef commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,124 @@
+	var hmacIdx = maxHops + position
+
+	// Iterate over all downstream nodes.
+	for j := 0; j < maxHops-position-1; j++ {
```
I find the inclusion of positional/index information here confusing: how will a forwarding node actually derive its position? If this is given to it by the sender, ofc that breaks w/e privacy we may have w/ the current scheme.

joostjager replied:
They do not need to derive their position. And this information also isn't given to them by the sender, because that would indeed break privacy.

The core idea of this proposal is to iterate over every possible position that the node may have in the path, and add a corresponding hmac for each of those positions.

Roasbeef commented on attributable_error_decrypt.go:
```diff
@@ -0,0 +1,207 @@
+		// Extract the payload and exit with a nil message if it is invalid.
+		source, payload, err := o.extractPayload(payloads)
+		if sender == 0 {
+			if err != nil {
```
So we don't handle the error at all if the sender is known?

joostjager replied:
The handling consists of returning the current position to the caller.

Some invalid data was encountered because we couldn't extract the payload. This means that there is no actual message to return. With the position information, the caller can apply a selective penalty.

Roasbeef reviewed (changes requested):
An initial pass to build up my mental model. Will sit with the spec for a bit before doing another pass here.

joostjager committed:
sphinx: add minPaddedOnionErrorLength constant

bitromortac reviewed (commented):
Nice work, quite a smart way to approach the problem!

Having `HopCount` is probably not ideal because it could be used for fingerprinting, but this ensures to have enough space for future extensions, right? This brings me to the question whether it's worthwhile to stay below the TCP MTU limit for latency reasons (similarly as for the onion size in `update_add_htlc`)?

If we have `b` bits (say 8000) available for the HMACs and security of `s` bits in the HMACs (HMAC chopped off after `s` bits, as you suggested out of band), we would get

`s = b / (HopCount * (HopCount + 1) / 2)`
```
HopCount |   s
--------------
27       |  21
20       |  38
18       |  47
16       |  59
14       |  76
12       | 102
```

For comparison, the Bitcoin network has a hash rate that manages s~80 in 10 min. In practice anything higher than that seems to be very high. The downside of having a fixed total message size below `b` is that the number of total hops is limited.

joostjager commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,138 @@
+
+// AttributableErrorStructure contains the parameters that define the structure
+// of the error message that is passed back.
+type AttributableErrorStructure struct {
```
Yes, `FixedPayloadLen` is to allow future flexibility without increasing the failure message until it is needed.

`HopCount` is slightly different. It gives senders some control over the failure message size, at the cost of limiting the maximum route length. If the size of the total error can be reduced so that `HopCount` can be hard-coded to the max route length of 27, that would be great.

joostjager commented on attributable_error_decrypt.go:
```diff
@@ -0,0 +1,209 @@
+
+	// Ensure the error message length is enough to contain the payloads and
+	// hmacs blocks. Otherwise blame the first hop.
+	if len(encryptedData) < minPaddedOnionErrorLength+o.hmacsAndPayloadsLen() {
```
This is good question. 

In master what happens is that a random failure message is made up if the length is below the (legacy) minimum:
https://github.com/lightningnetwork/lnd/blob/e549096b8837a95e0515c8c63fa01c634714d913/htlcswitch/link.go#LL1889C2-L1889C2

For attributable errors though, it may pass this check. Then in the intermediate EncryptError, an error is returned.

I think currently in https://github.com/lightningnetwork/lnd/pull/7139, the intermediate node ignores the error and passes back the failure message as is. This allows an attacker to get away with it.

What should happen instead is that the intermediate node creates a new, properly formatted failure to attribute blame to the connection between itself and its downstream peer.

joostjager commented:
>If we have b bits (say 8000) available for the HMACs and security of s bits in the HMACs (HMAC chopped off after s bits, as you suggested out of band)

Yes, I think this would be a useful optimization. You could argue that attributing failure isn't the most critical thing in lightning. Even with a relatively high chance of an attacker forging an hmac that is considered valid (let's say 21 bits, 1 in 2,000,000), game theory is still very much against them. If they guess wrong, they'll be penalized and will have to wait for the reputation to be restored before they'll receive new traffic from that sender.

joostjager commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,138 @@
+
+const (
+	// payloadIntermediateNode is a marker to signal that this attributable
+	// error payload is originating from a node between the payer and the error
```
Seems that my editor was set to tab width 4 for this repo. Fixed.

joostjager commented on attributable_error_crypto.go:
```diff
@@ -0,0 +1,138 @@
+	_, _ = hash.Write(message)
+
+	// Include payloads including our own.
+	_, _ = hash.Write(payloads[:(o.maxHops()-position)*o.payloadLen()])
```
Good catch, this parameter is indeed inverted. Updated the code to match the comment. In the spec pr, I also define position relative to the error soruce.

joostjager commented on testdata/attributable_error.json:
```diff
@@ -0,0 +1,25 @@
+{
```
This file is taken straight from the spec PR where the parameters are explained. I will do more polishing after we've decided that attr errors is going to happen and what variation of it exactly.

joostjager commented:
Added a parameter in the attributable error structure struct for the size of the hmac.

The other changes mentioned above can happen in the lnd pr.

bitromortac reviewed (approved):
LGTM :tada:! I compared the left/right-shift/HMAC operations to my notes and they are identical. Great to have the attributable errors completely parametrized such that we can define them for the protocol.

joostjager commented on attributable_error_decrypt.go:
```diff
@@ -0,0 +1,211 @@
+
+		return &DecryptedAttrError{
+			DecryptedError: DecryptedError{
+				SenderIdx: 1,
```
How would we do that? If our peer returns a message that is too short, then they didn't do what do should have done. They should always return a message of the correct length that includes at least their valid hmacs.

bitromortac replied:
I thought about distinguishing the happy path decoding (we can determine the true sender) from the unhappy one (`isCorruptedSender bool`), but I'm not sure how that would be used for penalization yet.

joostjager replied:
The unhappy path is currently signaled through a nil `DecryptedError.Message`. But perhaps it should be more explicit? This is where I miss rust's expressiveness in data types.

joostjager commented on attributable_error_crypto_test.go:
```diff
@@ -0,0 +1,432 @@
+
+// Generate a list of 4+3+2+1 = 10 unique hmacs. The length of this list is
+// fixed for the chosen maxHops.
+func createTestHmacs() []byte {
```
Yes good point. Updated.

joostjager commented:
Test vector updated to 20/4/4 structure in line with https://github.com/lightning/bolts/pull/1044#issuecomment-1669510607

GeorgeTsagk committed:
testdata: update vectors for attributable errors

We add the encoded failure message and the intermediate encrypted error
messages that correspond to attribution data to the test vector.

GeorgeTsagk committed:
sphinx: add attributable error structure

We add the AttrErrorStructure which is going to be used in follow-up
commits by the updated onion error encrypter and decrypter.

GeorgeTsagk committed:
sphinx: update legacy onion tests

Update existing onion error tests to now include the attribution data.
Since this is a backwards compatible change the legacy error message
should behave the same, so the asserts/vectors do not need to be updated.

GeorgeTsagk committed:
sphinx: add attributable error tests

We now add the attributable error test suite, which verifies that the
new encrypter/decrypter behave as expected against various attrData inputs.

joostjager commented:
>We could also "relax" the current decryption to not halt when running into invalid/empty attribution data, but optimistically continue decryption over the failure message.

This is what we do IN LDK. We always decrypt and use the attribution data to assign blame, if it is present. Once the network has been upgraded sufficiently, pathfinding can skip nodes that do not yet support it, or only use them as a last resort.

GeorgeTsagk committed:
sphinx: update encrypter for attr errors

We enhance the existing onion error encrypter to now include the
attribution data. This will be needed by upstream peers in order to
help verify the error source and hold times.

GeorgeTsagk committed:
sphinx: update decrypter for attr errors

We enhance the decrypter to now also use attr data. This structure is
now being decrypted alongside the legacy error message. If any of the
two fields fails to decrypt we stop the decryption process and blame the
further decryptable hop.

GeorgeTsagk committed:
sphinx: add strictAttribution flag for onion error decryption

We now add a flag to the decryption function which controls the way
we're carrying out the decryption with regards to the attribution data.
This is primarily done for graceful backwards compatibility, in order to
prevent nodes upgrading to attributable failures from blaming their
first hop peers, which would very fast blacklist their channels and
cause failed payments. As an initial step, we'd want the senders to
perform the non-strict decryption, eventually switching to the strict
decryption once adoption is sufficient.

GeorgeTsagk commented:
Rebased
